Create standard for mandatory and supported IaaS services

josephineSei opened the pull request:
closes https://github.com/SovereignCloudStack/issues/issues/528

markus-hentsch commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+
+## Supported OpenStack services
+
+The following services MAY be present in SCS-compliant IaaS-Deployment and they are considered in the SCS standards.
```
```suggestion
The following services MAY be present in SCS-compliant IaaS deployment and are considered in the SCS standards.
```

markus-hentsch commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+## Supported OpenStack services
+
+The following services MAY be present in SCS-compliant IaaS-Deployment and they are considered in the SCS standards.
+Most of these services (except Cloudkitty, Gnocchi and Masakari) have been integrated and tested by the SCS reference implementation:
```
```suggestion
```

This statement seems like a moving target to me, i.e. who is going to update this standard/statement if the component selection implemented in the reference setup changes?

I don't think this sentence adds any meaningful information to the standard that's why I'd remove it.

markus-hentsch commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+
+## Unsupported OpenStack services
+
+All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their integration and behavior by the SCS community.
```
```suggestion
All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their compatibility and behavior in SCS clouds by the SCS community.
```

markus-hentsch commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+## Unsupported OpenStack services
+
+All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their integration and behavior by the SCS community.
+Those services may be integrated into IaaS deployments by a CSP on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.
```
```suggestion
Those services MAY be integrated into IaaS deployments by a CSP on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.
```

markus-hentsch commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+## Conformance Tests
+
+The presence of the mandatory OpenStack services (except the S3) will be tested in a test-script.
+As S3 is a moving target, it may be integrated into the test, but will not let the Conformance test fail in general.
```
```suggestion
As the S3 interface is a moving target, it may be integrated into the test suite but the test result will not be taken into account to determine conformance.
```

markus-hentsch reviewed (commented):
Some suggestions for minor rephrasing of some things inline.

markus-hentsch commented:
Also one thing I only addressed in the review in a shallow manner so far:

I believe we discussed in the community calls previously that we need to make a clearer distinction between OpenStack APIs and the services and be careful about the wording in standards.

The OpenStack services are essentially reference implementations of the APIs. Wherever possible, the SCS should mandate the APIs and their functionalities, not the (backend) services. CSPs should be free to use whatever implementation they like as long as it provides the same API and behavior[^1]. Especially when mandating components, this can be a make or break deal for CSPs.


[^1]: not that I expect anybody to implement something like a full Nova alternative overnight but from a formal standpoint, the freedom should be there

anjastrunk commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+There are many OpenStack APIs and their corresponding services that can be deployed on the IaaS level.
+These services have differences in the quality of their implementation and liveness and some of them may be easily omitted when creating an IaaS deployment.
+To fulfill all SCS-provided standards only a subset of these APIs are required.
+Some more but not all remaining OpenStack APIs are also supported additionally by the SCS project and may be part of its reference implementation.
```
I struggle a bit with

> ... and may be part of its reference implementation

I though recommended services **are** part of SCS reference implementation.

josephineSei replied:
I think we still need to discuss this. Some services were requested, but currently are not present in the reference implementation (Cloudkitty, Heat, Masakari, Skyline). Heat might still be excluded from the supported list.

berendt replied:
Only Masakari is not available in the reference implementation at the moment. The other things are deployable, even if we would like to do without some of them. Masakari is also deployable, but we have not yet built and tested the necessary images.

artificial-intelligence replied:
Can this be marked as resolved?

anjastrunk commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+## Unsupported OpenStack services
+
+All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their compatibility and behavior in SCS clouds by the SCS community.
+Those services MAY be integrated into IaaS deployments by a CSP on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.
```
```suggestion
Those services MAY be integrated into IaaS deployments by a Cloud Service Provider on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.
```

anjastrunk reviewed (changes requested):
Two cosmetic issues and one question, which requires clarification.

artificial-intelligence commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+:::caution
+
+S3 API implementations may differ in certain offered features.
+CSPs must publicly describe, which implementation they use in their deployment.
```
Should we maybe describe in more detail how this should be documented?
Maybe we should just enforce a machine-readable format for that, given there are not that many implementations in practical usage (mainly swift and ceph-rgw)?

That would have the benefit that we can easily test the compliance of this part of the standard.

artificial-intelligence replied:
Apologies, I was indeed hinting at making the CSPs description machine readable, as I already thought that it might not be possible to do this via openstack API, I should have written that down more clearly.

artificial-intelligence replied:
I don't have an idea or proposal on how to actually implement this and as nobody else has picked this up I'm going to mark this as resolved. I still think it would be a good idea to make this easily testable.

artificial-intelligence commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,80 @@
+
+## Unsupported OpenStack services
+
+All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their compatibility and behavior in SCS clouds by the SCS community.
```
This wording could be read to imply that all supported services are tested for their compatibility and behavior, which I don't think is currently the case. If this is the case and I'm just unaware of the tests being done, please point me to these tests.

Also I'm not sure what is referred to when talking about "tested for their [..] behavior"? Does it mean the behavior of the service is SCS compliant - whatever that means, given that I'm not aware of any service specific standards with regards to the list of "supported" services? Maybe this sentence could be clarified a bit?

The only somewhat service specific standards I can think off of the top of my head are the image naming standard which could be considered to be glance specific and the entropy standard which could be viewed as nova specific, if you squint. So if there are service specific standards which we test against we should maybe list those here or at least link to them?

josephineSei replied:
In another comment @berendt mentioned that, except for Masakari, all supported services are available in the reference implementation. As I read his comment, they will eventually also test Masakari.

In deed there are standards that will and have to be applied to all used services: e.g. the role standard that is currently under development. This is also what I meant with behavior: if services have different role names or have a different meaning of role names, their behavior is different.
But i should rephrase this sentece, for better understanding.

artificial-intelligence replied:
Please don't conflate availability of services with "tested" services.
Not all services that are available to use are necessarily also tested.

Also I put tested in quotation marks, because it's really not defined anywhere what that actually means, beside SCS standard tests.

artificial-intelligence replied:
marking as resolved as we mostly talk about APIs now.

artificial-intelligence reviewed (changes requested):
the service lists themselves mostly LGTM, but I think some other parts could need some clarification, see the comments.

Thanks for working on this.

josephineSei commented:
I think we should discuss the supported services list in the IaaS call, as this topic is very important and should not be merged without having overall agreement also for the supported service list.

josephineSei commented:
We discussed this PR today in the IaaS Call. Especially what "supported service" means.

>   - What should users expect from these supported services?
>      - that standards refering to overall IaaS also include them (== all services can be enabled to fulfill the standards?)
>      - that these services have been tested for integration?
>      - that these services are part of the reference implementation? (imho that might be too much)
>          - "supported" in the standardization sense can not make statements on the reference implementation
>         - meaning of "supported" in standardization may be limited to ensuring that our standards don't conflict with these
>              - we don't break them with our standards!
>              - renaming "supported" to "recommended"? No: "recommended" has a stronger meaning ...
>              - we originally wanted to separate between 
>                  1. (mandatory) you need these APIs to have a scs-compliant cloud 
>                  2. (supported) you may use these APIs, they can be integrated with scs (e.g. work with the scs role standard)
>                  3. (unsupported) you may use these APIs on your own risk
>      - hints and recommendations for the implementation may be added to the implementation notes
> - Generic discussion: We do standardize concrete APIs (the ones that come from OpenStack or K8s or CAPI) to make the
> standards useful -- alternative implementations for services are possible, but a completely different set of technologies 
> will unlikely implement the SCS APIs.
> - We need to use the correct terminology "OpenStack Compute API" as opposed to "Nova"
>    - Complexity: These APIs are huge, with lots of optional and deprecated pieces
>        - Need to work through these, tedious
>        - The OpenStack Interop Guideline tests ("OpenStack powered XXX") are a good starting point for this
>            - This is a fairly small set, we need to add to this to produce something that is really useful ("this is the motivation for SCS standardization")

In conclusion the PR should focus mor on APIs than OpenStack services. Which has the downside, that we would also need to clarify for each API endpoint, whether this is mandatory, recommended or optional / not needed.

This is a direction which is worth going for, but this will add a load of complexity and might need specific documents for each service API.

artificial-intelligence commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,77 @@
+To fulfill all SCS-provided standards only a subset of these APIs are required.
+Some more but not all remaining OpenStack APIs are also supported additionally by the SCS project and may be part of its reference implementation.
+This results in different levels of support for specific services.
+This document will give readers insight about how the SCS classifies the OpenStack services accordingly.
```
not sure, but I guess we should also write APIs here?

```suggestion
This document will give readers insight about how the SCS classifies the OpenStack APIs accordingly.
```

artificial-intelligence commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,77 @@
+This results in different levels of support for specific services.
+This document will give readers insight about how the SCS classifies the OpenStack services accordingly.
+If a cloud provides all mandatory and any number of supported OpenStack APIs, it can be tested for SCS-compliance.
+Any unsupported services will not be tested.
```
```suggestion
Any unsupported APIs will not be tested.
```

artificial-intelligence commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,77 @@
+
+All other OpenStack services, which APIs are not mentioned in the mandatory or supported lists will not be tested for their compatibility and conformance in SCS clouds by the SCS community.
+Those services MAY be integrated into IaaS deployments by a Cloud Service Provider on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.
+The SCS standard offers no guarantees for compatibility or reliability of services categorized as unsupported in conjunction with an SCS-conformant infrastructure.
```
just personal opinion, I would shorten that sentence to make it easier to understand/read:

```suggestion
The SCS standard offers no guarantees for compatibility or reliability of services categorized as unsupported.
```

artificial-intelligence reviewed (changes requested):
I think I found mostly some typos and some missing replacements of /services/APIs/.

josephineSei commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -46,27 +45,25 @@ Users should always research whether a needed feature is supported in the offere
+## Supported IaaS APIs
 
-The following services MAY be present in SCS-compliant IaaS deployment and are considered in the SCS standards.
+The following IaaS APIs MAY be present in SCS-compliant IaaS deployment, e.g. implemented thorugh the corresponding OpenStack services, and are considered in the SCS standards.
```
No these APIs are truely optional, but in contrast to the unsupported APIs, we know they are likely to be used and such consider them in standards and (except for Masakari, if I understood @berendt correctly) even tested for integration.

I can edit the sentece to better reflect the wide but optional usage of these APIs, which is the reason for us to also think of them in standards and maybe sometimes test their integration?

josephineSei replied:
Is it because of the last part of the sentence? "... are considered in the SCS standards"? This is, because we need (for some standards) to consider MANY possible cloud configurations. The role standard for example with the new RBAC roles must decide between old and new roles, which has to be rolled out throughout all services in a deployment. If at some point we want to use the new RBAC system, all servies need to be able to do this, and while the mandatory service already are in a good state, some supported services are not. SO we do have to consider all only supported services in this standard, because many clouds use them, and we don't want to break them.
The mandatory Services are more of a minimal list of services, that are needed to have an scs-compliant cloud.

I hope this explains it a little bit better.

artificial-intelligence replied:
No, sorry, I think I didn't make clear what I meant:

In line 50 we have:

```suggestion
The following IaaS APIs MAY be present [..]
```

but in line 52, immediately following this we have:

```
| **Mandatory** API | corresponding OpenStack Service | description |
```

which imho is a contradiction in the heading.

josephineSei replied:
Uh I made a mistake when adding the first column to the table. It should be "Supported API". Thank you for pointing it out, I changed it.

artificial-intelligence commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,77 @@
+
+The following IaaS APIs MAY be present in SCS-compliant IaaS deployment, e.g. implemented thorugh the corresponding OpenStack services, and are considered in the SCS standards.
+
+| Mandatory API | corresponding OpenStack Service | description |
```
see discussion above

```suggestion
| API | corresponding OpenStack Service | description |
```

josephineSei replied:
I changed it to "Supported API"

josephineSei commented:
When I'm correct the only thing missing here for this standard is:
Is the Heat API still in the supported list or not?
@garloff and @fkr you wanted to discuss this, right? Do you have any news about it?

markus-hentsch commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,77 @@
+|-----|-----|-----|
+| **key-manager** | Barbican | Key Manager service |
+| **billing** | Cloudkitty | Rating/Billing service |
+| **telemetry** | Ceilomete | Telemetry service |
```
```suggestion
| **telemetry** | Ceilometer | Telemetry service |
```

mbuechse commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,77 @@
+## Introduction
+
+To be SCS-compliant a Cloud Service Provider (CSP) has to fulfill all SCS standards.
+Some of those standards are broad and consider all APIs of all services on the IaaS-Layer.
```
Can you please be specific and name at least one example?

gtema commented on Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md:
```diff
@@ -0,0 +1,77 @@
+| **image** | Glance | Image service |
+| **load-balancer** | Octavia | Load-balancer service |
+| **network** | Neutron | Networking service |
+| **s3** or **object-store** | S3 API object storage | No formal standard exists, many implementations: Swift, RadosGW, minio... |
```
I personally dislike "or" here. There are user loads supporting only one of both so having here an OR does not bring much benefit. S3 support in it's own is not standardized enough to have multiple compatibility issues, especially when it comes to the OpenStack Identity related stuff.

gtema replied:
We discussed this in the standardization SIG and following items were raised:
- S3 was always intended to be mandatory (also for KaaS layer)
- Swift should be at least "recommended"
Even though S3 and Swift both implement the same service (object-store) it can be seen as: HDMI output is a must and DisplayPort is recommended

mbuechse replied:
And it should be two separate rows

josephineSei replied:
I will change the standard accordingly. 
**BUT**, if S3 is also listed as "object-store" equal to Swift, we have no chance to distinguish between those two services in the tests. 
This is a little bit unfortunate, because we will never know for certain, whether the endpoints are from an S3 service or from Swift. So I can change the lines here, but we will not be able to test them!

An option may be to use gaia-x credentials. Which raises the question: Do we have a point, were we state which credentials we require? @garloff ?

anjastrunk replied:
IMO, in this context, we need a standard for object storage too, similar like for block storage. CSPs need to know the requirements for object store and users need to know, which capabilities an object store has. See, #725

anjastrunk replied:
I agree to @josephineSei: We can not distinguish between two object store APIs in tests. We should define just one as mandatory.  Furthermore, setting S3 as mandatory is  problematic. As far as I known, S3 API is not in control of OpenStack. There is no official documentation on how S3 for a specific OpenStack release looks like. The latest documentation is from [Mitaka](https://docs.openstack.org/mitaka/config-reference/object-storage/configure-s3.html).  There is no documentation for (current release)[https://docs.openstack.org/2024.1/config-reference/object-storage/configure-s3.html], e.g. Hence, we standardize a moving target IMO.

anjastrunk replied:
There is still an open issue  with respect to version of a certain OpenStack service.  E.g. how does a consumer knows, which version of Nova is provided by a CSP? We will answer this question in #765

gtema commented on Tests/iaas/mandatory-services/mandatory-iaas-services.py:
```diff
@@ -0,0 +1,96 @@
+logger = logging.getLogger(__name__)
+mandatory_services = ["compute", "identity", "image", "block-storage",
+                      "network", "load-balancer", "placement"]
+object_store_service = ["s3", "object-store"]
```
I have never seen a single cloud having S3 service/endpoint in the catalog. Typically `object-store` is also supporting S3 (either it is swift with S3 enabled or it is rgw catalog pointing to swift endpoint).

gtema replied:
There is no tooling existing in the world (known to me) that expects S3 and consumes OpenStack service catalog searching for "s3" service type

gtema reviewed (changes requested):
S3 with/vs Swift requires a bit more precision in my eyes

anjastrunk commented:
An other question came to my mind: How does a user know, which API version a certain SCS cloud supports? There is a [document on certificates](https://docs.scs.community/standards/scs-compatible-iaas) listing all effective standards for a certain scope and version. Standard on mandatory services will be part of a scopes/version, soon. How does the user/reader figure out, which API of which OpenStack service belongs to a given certificate scope and version.

josephineSei commented:
As discussed in the IaaS meeting, we want to standardize, that the s3 should be mandatory and swift only supported.
I looking into the s3 protocol and the hints for tests from Kurt to adjust the current testing behavior.

The one open question to me (which needs to be tested) is how do we get an endpoint for s3 when we do not have swift? or do we require that swift should be always there - even just as a stump to provide that endpoint in the service catalog?

josephineSei commented:
I was able to test my script and Kurts script against a deployment from Cloud&Heat, which provides a Swift endpoint and also offers S3 behind this endpoint.

But it also came up, that there was another description for "block-storage" that was just volume. I need to adjust the test, to allow both options - or do we want to only allow one specific endpoint name:
1. do we require that all endpoints have to show up in `catalog list` output?
2.  do we want to maybe only allow one name per service (only block-storage OR volume for the block storage service)?

I would like to answer the first question with yes, because in this way we will always have access to the object storage endpoint, regardless, what service is used.

The second question is more difficult: we may need to adjust the tests to allow more than one name per service - each time a new name comes up. On the other hand: that will have a huge impact on CSPs, so I don't think, we should strictly only allow one name.

josephineSei commented:
I installed minio in a test environment and started it temporarily:
```
root@devstack:/opt/stack/devstack# ./minio server /data
MinIO Object Storage Server
Copyright: 2015-2024 MinIO, Inc.
License: GNU AGPLv3 - https://www.gnu.org/licenses/agpl-3.0.html
Version: RELEASE.2024-09-22T00-33-43Z (go1.22.7 linux/amd64)

API: http://192.168.23.161:9000  http://192.168.11.28:9000  http://10.0.2.1:9000  http://192.168.122.1:9000  http://127.0.0.1:9000 
   RootUser: minioadmin 
   RootPass: minioadmin 

WebUI: http://192.168.23.161:46367 http://192.168.11.28:46367 http://10.0.2.1:46367 http://192.168.122.1:46367 http://127.0.0.1:46367          
   RootUser: minioadmin 
   RootPass: minioadmin 

CLI: https://min.io/docs/minio/linux/reference/minio-mc.html#quickstart
   $ mc alias set 'myminio' 'http://192.168.23.161:9000' 'minioadmin' 'minioadmin'

Docs: https://docs.min.io
WARN: Detected default credentials 'minioadmin:minioadmin', we recommend that you change these values with 'MINIO_ROOT_USER' and 'MINIO_ROOT_PASSWORD' environment variables

```

In another console tab I installed the minio client, logged in as admin, created a new user and gave that users rights to read and write:
```
stack@devstack:~/devstack$ curl https://dl.min.io/client/mc/release/linux-amd64/mc --create-dirs -o $HOME/minio-binaries/mc
stack@devstack:~/devstack$ mc alias set 'myminio' 'http://192.168.23.161:9000' 'minioadmin' 'minioadmin'
Command 'mc' not found, but can be installed with:
apt install mc
Please ask your administrator.
stack@devstack:~/devstack$ chmod +x $HOME/minio-binaries/mc
stack@devstack:~/devstack$ export PATH=$PATH:$HOME/minio-binaries/
stack@devstack:~/devstack$ mc alias set 'myminio' 'http://192.168.23.161:9000' 'minioadmin' 'minioadmin'
Added `myminio` successfully.
stack@devstack:~/devstack$ mc admin user add myminio newuser newusersecret
Added user `newuser` successfully.
stack@devstack:~/devstack$ mc admin user list myminio
enabled    newuser                                   
stack@devstack:~/devstack$ mc admin policy list myminio
consoleAdmin
diagnostics
readonly
readwrite
writeonly
stack@devstack:~/devstack$ mc admin policy attach myminio readwrite --user newuser
Attached Policies: [readwrite]
To User: newuser
```

After this I was able to test the script and made adjustments to it until I reached the point, where it succeeded:
```
stack@devstack:~/devstack$ python3 mandatory-iaas-services3.py --os-cloud mycloud2 --s3-endpoint "http://192.168.23.161:9000" --s3-access newuser --s3-access-secret newusersecret
FAIL: The following endpoints are missing: ['load-balancer']
SUCCESS: S3 exists
```

When not giving the parameters it will fail:
```
stack@devstack:~/devstack$ python3 mandatory-iaas-services3.py --os-cloud mycloud2
FAIL: The following endpoints are missing: ['load-balancer', 'object-store']
FAIL: No object store endpoint found. No testing for the s3 service possible. Details: public endpoint for object-store service not found
```

josephineSei commented:
@gtema could you review this standard again?

gtema commented on Tests/iaas/mandatory-services/mandatory-iaas-services.py:
```diff
@@ -0,0 +1,299 @@
+    s3_from_ostack(s3_creds, connection, endpoint)
+    # Overrides (var names are from libs3, in case you wonder)
+    s3_from_env(s3_creds, "HOST", "S3_HOSTNAME", "https://")
+    s3_from_env(s3_creds, "AKI", "S3_ACCESS_KEY_ID")
```
The most common way to name those is AK/SK which is immediately understood by anybody used S3 at least one time in life. AKI/SAK introduces new not very clear naming scheme

josephineSei replied:
I changed this.

josephineSei commented:
@fkr or @garloff do you want to take another look at this?
